### PR TITLE
add ArrayAccess support for BaseArrayHelper

### PR DESCRIPTION
**Pull request by mikk150**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌
| Tests pass?   | ✔️
| Fixed issues  | #18026

### CODE REVIEW

File: framework/helpers/BaseArrayHelper.php [hunk 1]
```diff
@@ -849,6 +861,21 @@ public static function isTraversable($var)
...
     }
 
+    /**
+     * Checks whether a variable is an array or [[\ArrayAccess]].
+     *
+     * This method does the same as the PHP function [is_array()](https://secure.php.net/manual/en/function.is-array.php)
+     * but additionally works on objects that implement the [[\ArrayAccess]] interface.
+     * @param mixed $var The variable being evaluated.
+     * @return bool whether data on $var can be accessed as arrays
+     * @see https://secure.php.net/manual/en/function.is-array.php
```
Comment (samdark) [inline framework/helpers/BaseArrayHelper.php:871]: Need `@since 2.0.36`

File: framework/helpers/BaseArrayHelper.php [hunk 2]
```diff
@@ -203,17 +204,28 @@ public static function getValue($array, $key, $default = null)
...
             // it is not reliably possible to check whether a property is accessible beforehand
             return $array->$key;
-        } elseif (is_array($array)) {
-            return (isset($array[$key]) || array_key_exists($key, $array)) ? $array[$key] : $default;
         }
 
         return $default;
     }
 
+    private static function arrayHasKey($array, $key)
```
Comment (samdark) [inline framework/helpers/BaseArrayHelper.php:218]: Does it worth making it public?

Comment (mikk150) [reply framework/helpers/BaseArrayHelper.php:218]: yes, I can actually make it public if you want

Comment (samdark) [reply framework/helpers/BaseArrayHelper.php:218]: Well... if you find it useful.

Comment (mikk150) [reply framework/helpers/BaseArrayHelper.php:218]: I think I will much rather fix `BaseArrayHelper::keyExists()` method that this class already has...

File: framework/helpers/BaseArrayHelper.php [hunk 3]
```diff
@@ -602,7 +604,18 @@ public static function keyExists($key, $array, $caseSensitive = true)
         if ($caseSensitive) {
             // Function `isset` checks key faster but skips `null`, `array_key_exists` handles this case
             // https://secure.php.net/manual/en/function.array-key-exists.php#107786
-            return isset($array[$key]) || array_key_exists($key, $array);
+            if (is_array($array) && (isset($array[$key]) || array_key_exists($key, $array))) {
+                return true;
+            }
+            // Cannot use `array_has_key` on Objects for PHP 7.4+, therefore we need to check using [[ArrayAccess::offsetExists()]]
+            if ($array instanceof ArrayAccess && $array->offsetExists($key)) {
```
Comment (samdark) [inline framework/helpers/BaseArrayHelper.php:611]:
```suggestion
            return $array instanceof ArrayAccess && $array->offsetExists($key));
```

File: framework/helpers/BaseArrayHelper.php [hunk 4]
```diff
@@ -203,12 +205,12 @@ public static function getValue($array, $key, $default = null)
             $key = substr($key, $pos + 1);
         }
 
-        if (is_object($array)) {
+        if (static::isArrayAccess($array)) {
+            return static::keyExists($key, $array) ? $array[$key] : $default;
+        } elseif (is_object($array)) {
```
Comment (samdark) [inline framework/helpers/BaseArrayHelper.php:210]:
```suggestion

        if (is_object($array)) {
```

**Review by samdark — APPROVED** [review]
Looks good overall 👍

**Comment by samdark** [comment]
Thanks!

**Pull request merged**